### PR TITLE
Add RTR flags to fix sync race in PRK Transpose

### PR DESCRIPTION
**Pull request by jdinan**
Fix a race in the iterations loop where iterations can overlap resulting in data corruption and deadlock.  The fix adds another set of flags to inform the sender that the receiver is ready to receive the message.

Note: this fix will have a negative impact on performance of the flag update is blocking in the SHMEM library.

### CODE REVIEW

File: SHMEM/Transpose/transpose.c [hunk 1]
```diff
@@ -370,6 +378,9 @@ int main(int argc, char ** argv)
       for (j=0; j<Block_order; j++)
         for (i=0; i<Block_order; i++) 
           B(i,j) += Work_in(phase, i,j);
+
+      /* Tell sender that we are ready to for the next iteration */
+      shmem_int_inc(&send_flag[my_ID], recv_from);
```
Comment (jdinan) [inline SHMEM/Transpose/transpose.c]: Either ```shmem_int_p``` or ```shmem_int_inc``` can be used here.  Do you have a preference?

**Comment by jdinan** [comment]
@rfvander @jeffhammond Could you take a look at this?

**Comment by rfvander** [comment]
There are several failures. It looks like they don’t have anything to do with your change. I noticed a problem with OPENMP/Refcount, which relies on a hint for how to deal with lock access, specifically: omp_init_lock_with_hint(pcounter_lock,omp_lock_hint_uncontended);
Apaprently, this is Intel-only at present. I’ll need to do a conditional compile for this option.
Another error has to do with how to pass options in BUPC to the compiler.
I’ll fix these problems, but it may take a little while, as I am racing to complete a Charm++ study.

Rob

**Comment by jdinan** [comment]
I don't see how those failures could be a result of a change to the SHMEM ```transpose.c``` file.

This bug is unfortunately wreaking havoc on the SOS CI testing.  I'll point SOS at my PRK fork until you're able to get to this.

**Comment by rfvander** [comment]
I agree, Jim, I just looked at the first set of failed tests, and my observations are about those. Sorry.

**Comment by jdinan** [comment]
Bump.

**Comment by rfvander** [comment]
Done

**Comment by rfvander** [comment]
Um, I hope I did the right thing with the pull request.

**Comment by jdinan** [comment]
Nope, you did not.  :)

I reopened the PR.  Please review the code (and the comment I left above on atomic_inc vs put).  If it looks good, merge.  If not, let me know what you would like me to fix.

**Comment by jeffhammond** [comment]
Put is always faster than increment in hardware...

**Comment by jdinan** [comment]
Updated.

**Comment by jdinan** [comment]
@rfvander The update looks good.  One suggestion is that you might also want to add the BARRIER_SYNCH=1 build to Travis.

**Pull request merged**